src/Symfony/Component/Uri/Uri.php

```diff
@@ -0,0 +1,130 @@
+ *
+ * @see https://tools.ietf.org/html/rfc3986
+ */
+final class Uri implements \Stringable
+{
+    private const URI_GLOBAL_REGEX =
+        '/^(?:(?P<scheme>[^:\/?#]+):)?(?:\/\/(?P<authority>[^\/?#]*))?(?P<path>[^?#]*)(?:\?(?P<query>[^#]*))?(?:#(?P<fragment>.*))?$/';
+
+    private const URI_AUTHORITY_REGEX = '/^(?:(?P<user>[^:@]*)(?::(?P<pass>[^@]*))?@)?(?P<host>[^:]*)(?::(?P<port>\d*))?$/';
+
+    public function __construct(
+        public string $scheme,
+        public ?string $user = null,
+        public ?string $password = null,
+        public ?string $host = null,
+        public ?int $port = null,
+        public ?string $path = null,
+        public ?QueryString $query = null,
+        public ?string $fragment = null,
+    ) {
+    }
+
+    /**
+     * Parses a URL.
+     *
+     * The `user` and `pass` keys are url-decoded automatically when parsing.
+     *
+     * @throws InvalidUriException
+     */
+    public static function parse(string $uri): static
+    {
+        preg_match(self::URI_GLOBAL_REGEX, $uri, $matches);
+        if (!$matches || !isset($matches['scheme']) || '' === $matches['scheme']) {
+            throw new InvalidUriException($uri);
+        }
+
+        if (isset($matches['authority'])) {
+            preg_match(self::URI_AUTHORITY_REGEX, $matches['authority'], $authMatches);
+
+            $matches = array_merge($matches, $authMatches);
+            unset($matches['authority']);
+        }
+
+        $matches = array_filter($matches, static fn (string $value): bool => $value !== '');
+
+        return new static(
+            $matches['scheme'],
+            isset($matches['user']) ? urldecode($matches['user']) : null,
+            isset($matches['pass']) ? urldecode($matches['pass']) : null,
+            $matches['host'] ?? null,
+            $matches['port'] ?? null,
+            $matches['path'] ?? null,
+            isset($matches['query']) ? QueryString::parse($matches['query']) : null,
+            $matches['fragment'] ?? null,
+        );
+    }
+
+    /**
+     * A few examples of input and output:
+     *
+     * Uri::resolve('http://example.com', '/foo/bar'); // http://example.com/foo/bar
+     * Uri::resolve('http://example.com/foo', '/bar'); // http://example.com/bar
+     * Uri::resolve('http://example.com/foo', 'bar'); // http://example.com/foo/bar
+     * Uri::resolve('http://example.com/foo/', 'bar'); // http://example.com/foo/bar
+     * Uri::resolve('http://example.com/foo/', '../bar'); // http://example.com/bar
+     * Uri::resolve('http://example.com/foo/', '../../bar'); // http://example.com/bar
+     * Uri::resolve('http://example.com/foo/', '/bar'); // http://example.com/bar
+     * Uri::resolve('http://example.com/foo/', 'http://example.org/bar'); // http://example.org/bar
+     */
+    public static function resolve(Uri|string $baseUri, string $relativeUri): string
+    {
+        $baseUri = $baseUri instanceof Uri ? $baseUri : self::parse($baseUri);
```

Comment:
Micro optimisation: Move this down to line 99.

Reply:
Good idea